src/network/stream_reader.rs

```diff
@@ -0,0 +1,310 @@
+                0xff, 0x7f, 0x00, 0xff, 0xff, 0xff, 0x7f, 0x00,
+                0x2f, 0x55, 0x52, 0x47, 0x45, 0x4e, 0x54, 0x3a,
+                0x20, 0x41, 0x6c, 0x65, 0x72, 0x74, 0x20, 0x6b,
+                0x65, 0x79, 0x20, 0x63, 0x6f, 0x6d, 0x70, 0x72,
+                0x6f, 0x6d, 0x69, 0x73, 0x65, 0x64, 0x2c, 0x20,
+                0x75, 0x70, 0x67, 0x72, 0x61, 0x64, 0x65, 0x20,
+                0x72, 0x65, 0x71, 0x75, 0x69, 0x72, 0x65, 0x64,
+                0x00,
+            ].to_vec());
+        } else {
+            panic!("Wrong message type: expected AlertMessage");
+        }
+    }
+
+    #[test]
+    fn parse_multipartmsg_test() {
+        let mut tmpfile: File = tempfile::tempfile().unwrap();
+        let mut reader = StreamReader::new(&mut tmpfile, None);
+        reader.unparsed = MSG_ALERT[..24].to_vec();
+        assert!(reader.next_message().is_err());
+        assert_eq!(reader.unparsed.len(), 24);
+
+        reader.unparsed = MSG_ALERT.to_vec();
+        let message = reader.next_message().unwrap();
+        assert_eq!(reader.unparsed.len(), 0);
+
+        check_alert_msg(&message);
+    }
+
+    fn init_stream(buf: &[u8]) -> File {
+        let mut tmpfile: File = tempfile::tempfile().unwrap();
+        write_file(&mut tmpfile, &buf);
+        tmpfile
+    }
+
+    fn write_file(tmpfile: &mut File, buf: &[u8]) {
+        tmpfile.seek(SeekFrom::End(0)).unwrap();
+        tmpfile.write(&buf).unwrap();
+        tmpfile.flush().unwrap();
+        tmpfile.seek(SeekFrom::Start(0)).unwrap();
+    }
+
+    #[test]
+    fn read_singlemsg_test() {
+        let mut stream = init_stream(&MSG_VERSION);
+        let message = StreamReader::new(&mut stream, None).next_message().unwrap();
+
+        check_version_msg(&message);
+    }
+
+    #[test]
+    fn read_doublemsgs_test() {
+        let mut stream = init_stream(&MSG_VERSION);
+        write_file(&mut stream, &MSG_PING);
+
+        let mut reader = StreamReader::new(&mut stream, None);
+        let message = reader.next_message().unwrap();
+        check_version_msg(&message);
+
+        let msg = reader.next_message().unwrap();
+        assert_eq!(msg.magic, 0xd9b4bef9);
+        if let NetworkMessage::Ping(nonce) = msg.payload {
+            assert_eq!(nonce, 100);
+        } else {
+            panic!("Wrong message type, expected PingMessage");
+        }
+    }
+
+    fn serve_tcp(port: u16, pieces: Vec<Vec<u8>>) {
+        let listener = TcpListener::bind(format!("127.0.0.1:{}", port)).unwrap();
+        for ostream in listener.incoming() {
+            let mut ostream = ostream.unwrap();
+
+            for piece in pieces {
+                ostream.write(&piece[..]).unwrap();
+                ostream.flush().unwrap();
+                thread::sleep(Duration::from_secs(1));
+            }
+
+            ostream.shutdown(Shutdown::Both).unwrap();
+            break;
+        }
+    }
+
+    #[test]
+    fn read_multipartmsg_test() {
+        let port: u16 = 34254;
```

Comment:
We can use a port number of 0 to have the OS assign a port. See: https://doc.rust-lang.org/std/net/struct.TcpListener.html#impl

Reply:
Done